types/src/diem_id_identifier.rs

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{de, Serialize};
+use std::{
+    fmt::{Display, Formatter},
+    str::FromStr,
+};
+
+/// This file implements length and character set limited string types needed
+/// for DiemID as defined by DIP-10: https://github.com/diem/dip/blob/main/dips/dip-10.md
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DiemId {
+    // A reusable identifier that represents either the source or destination
+    // end-user of a payment transaction. It is unique to per user at VASP level
+    user_identifier: DiemIdUserIdentifier,
+    // A unique string that is mapped to a VASP
+    vasp_domain_identifier: DiemIdVaspDomainIdentifier,
+}
+
+impl DiemId {
+    pub fn new(
+        user_identifier: DiemIdUserIdentifier,
+        vasp_domain_identifier: DiemIdVaspDomainIdentifier,
+    ) -> Self {
+        Self {
+            user_identifier,
+            vasp_domain_identifier,
+        }
+    }
+
+    fn new_from_raw(
+        user_identifier: &str,
+        vasp_domain_identifier: &str,
+    ) -> Result<Self, DiemIdParseError> {
+        Ok(DiemId::new(
+            DiemIdUserIdentifier::new(user_identifier)?,
+            DiemIdVaspDomainIdentifier::new(vasp_domain_identifier)?,
+        ))
+    }
+
+    pub fn user_identifier(&self) -> &DiemIdUserIdentifier {
+        &self.user_identifier
+    }
+
+    pub fn vasp_domain_identifier(&self) -> &DiemIdVaspDomainIdentifier {
+        &self.vasp_domain_identifier
+    }
+}
+
+impl Display for DiemId {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{}@{}",
+            self.user_identifier.as_str(),
+            self.vasp_domain_identifier.as_str()
+        )
+    }
+}
+
+impl FromStr for DiemId {
+    type Err = DiemIdParseError;
+
+    fn from_str(s: &str) -> Result<Self, DiemIdParseError> {
+        let index = s
+            .find('@')
+            .ok_or_else(|| DiemIdParseError::new("DiemId does not have @".to_string()))?;
+        let split = s.split_at(index);
+        DiemId::new_from_raw(split.0, (split.1).trim_start_matches('@'))
```

Comment:
The `trim_start_matches` function results in `bmwill@@novi` being able to pass the validation. Lets instead change this to 
```suggestion
        DiemId::new_from_raw(split.0, &(split.1)[1..])
```

Reply:
Lets also add a test case for this example